### DISCUSSION TITLE
Is Supabase Truly Production Ready?

### DISCUSSION THREAD

**Discussion by terehov** [supabase/supabase#28377] [Questions] [answered]
We are evaluating Supabase for our next project, and so far we like it quite a bit. However, I have some concerns about whether Supabase is truly production-ready. I know it's claimed to be, but I have some doubts:

1. **Backups**: 
   - The backups seem to only concern the database. Buckets are not backed up.
   - Files are deleted immediately, and if a database backup is restored, the entries in the storage schema no longer match the buckets.
   - There's no real way to back up the buckets either. The S3 compatibility is far from being complete and not suitable for this purpose.
   - The recommendation to bring your own storage provider is also not helpful since this functionality is also not complete.
   - How can one establish a proper backup strategy for a core feature in a production environment?

2. **Row Level Permissions**:
   - Supabase is built on PostgreSQL and uses Row Level Permissions, which is great.
   - However, the permissions are not included in the migrations.

3. **Branching**:
   - Branching does not seem to be fully implemented. It appears to set up a new database and apply the migrations and seeds.
   - Permissions are not transferred, and there's no way to promote versions.

Am I overlooking something? Is there a recommended approach to address these concerns? Any guidance or clarification would be greatly appreciated. Thank you!

**Comment by gc-ft** [comment, marked as answer]
Hi @terehov,

**1. Backups**
I totally agree with your Backups concern related to buckets. A good system for backing up storage is missing clearly. In my opinion the storage system of supabase needs to be used in connection with your own management of files, meaning you need to keep records of your files on your own on top of the storage schema, considering the storage schema in the db itself is mainly just for supabase to know what files it has in storage and to allow for RLS policies to work with files.

So if you do not want files deleted right away, simply do not use the storage system directly, instead flag files on your own end as "deleted" so you no longer show them, and use a cron or the like to flush "the bin" manually when needed. That's how I would "solve" the "files are deleted immediately" sync issue...

This of course does not solve backing up the buckets somewhere themselves for safety, however I guess replicating the files to your own S3 should not be a very hard setup using a trigger?

**2. RLS**
Row Level Security Policies are in migrations, not sure why you believe this is not the case? The migrations have CREATE POLICY/ lines in them. If you are missing policies from schemas that are not "public" or a schema you added manually, that is because only the schemas "extensions", "public" and any custom schema you created are placed in migrations by default. So you have to manually create migrations for other schemas as needed, usually storage and/or auth.

When you create a migration via `supabase db pull` it actually warns you about this:
```
The auth and storage schemas are excluded. Run supabase db pull --schema auth,storage again to diff them.
```

**3. Branching**
I understand why you claim permissions are not transferred if you have the issue of 2. - But I do not see how this can be the case, there must be something else going on at your end that is causing this to fail (again, unless we are talking about storage schema or such).

Branching is not meant in the way you seem to want to use it however! It is more of a "development" branching and not a "staging" logic.. Because data is not copied (just seeds), it makes little sense in my opinion to offer to "promote" the branch to production? I believe the intended purpose of this is not to have a fully working copy of the supabase instance, just a quick copy to run tests against and do some final testing before the next step.

I have been thinking about CI/CD pipeline logic with supabase the last days myself, and the best setup for me is the following:

1) Two supabase projects, one for production, one for staging.
2) Use GitHub Actions, instead of the Supabase <> Github Integration
3) Best GitHub Workflows logic for me would be:
- Any pull request opened for develop branch in git, should create a supabase branch via terraform in the staging project, to run tests and see if everything works in general. This supabase branch can be removed when the pull request is closed.
- Any push onto develop branch in git runs tests and if ok will push new migrations and all functions to supabase staging project
- Any pull request opened on main branch in git, should again create a supabase branch but this time on the production project, possibly with more complete seed data and/or other things to be tested that only make sense to test prior to production release.
- Any push onto main branch in git pushes the new migrations and functions to the supabase production project

There is a good example for Github workflows working this way here: [https://github.com/supabase/supabase-action-example](https://github.com/supabase/supabase-action-example)
This is just missing the functions part! There is an example for this in the Edge Functions Docs though: [https://supabase.com/docs/guides/functions/cicd-workflow](https://supabase.com/docs/guides/functions/cicd-workflow)

Could be this is possible with a mix of supabase <> GitHub integration AND GitHub Actions, but considering there is terraform, I am not sure if it is really needed. The only additional benefit of using GitHub integration for branches would be the integration with Vercel if you use Vercel too.

Fabian

**Comment by dunivlorraine** [comment]
Short answer (I'll detail it later):
- the cloud version seems to be close to be production-ready
- the self-hosted one is just barely usable as a development environment

**Reply by giosueDelgado** [reply]
Can you explain better?

**Reply by dunivlorraine** [reply]
Hi!

First of all, a little disclaimer: I have used Supabase for an enterprise project. We used the self-hosted option, as the Cloud one doesn't fit our requirements. That was between September to December 2024, and things may have evolved since that time (I'll try to take that into account).

Here's what I had to do in order to achieve a production deployment (and oh god, that wasn't fun).

First, because there is [no generic OAuth provider](https://github.com/orgs/supabase/discussions/6547) (and I couldn't setup a Keycloak as an intermediary), I had to fork the [auth repository](https://github.com/supabase/auth) and add our provider myself. This is then rebuilt using Docker and used as an image in the final compose stack. I also noted that the URLs and endpoints are [hardcoded](https://github.com/supabase/auth/blob/master/internal/api/provider/keycloak.go) and don't use the well-known file, that could saves me a lot of time because it would means that changing the well-known file could have suffice.

In top of that, the dev experience feels unpleasant if you have any custom service (e.g. our auth service). This is because `supabase` command that supports it needs the [ the `--db-url` option](https://supabase.com/docs/reference/cli/supabase-db-pull) every time (that's not the case everywhere) if you need to interact with your local base. Indeed, the `supabase start` command launches services that cannot be changed (image versions specifically). One could argue that [supabase link](https://supabase.com/docs/reference/cli/supabase-link) [allows you to do that](https://github.com/supabase/cli/issues/2391), but I've never found a way to link my local dev base to our self-hosted one. This leads to two different methods:
- you create a compose stack with different image versions but it's a bummer to sync with other devs;
- you use the CLI to streamline your dev workflow but it's not what your production stack looks like.

The second one seems fine as long as you enforce versions, and even so, sometimes, things just broke unexpectedly (the Vector service just broke without any reason in a latest update, and we had to remove it from the stack manually with the `-x` option).

About migrations: you can easily create using the CLI, but to effectively apply them, you have to create your own script/automated service. Why? Again, the Supabase linking cannot accept a self-hosted instance. So, if you don't want to expose your database port and URL to the world and want a nice CI/CD workflow, you have to create a small image that will use the CLI to push these migrations up to the remote Supabase instance within the same stack (or something like that). I don't blame Supabase for that, it's pretty common to do that with other databases, but the fact that the command is already there and ready to be used, but only for Cloud instances, doesn't seems right.

Regarding the UI: it's nice! It's sometimes hard to find what you want, but there are a lot of options so you don't always have to use the CLI. However, the displayed errors are not very useful and generic. API errors are thrown for literally anything (including database errors).

One big thing for us that wasn't really obvious is the auth schema. First, it cannot be used to trigger functions easily (you can't add functions or triggers to the auth schema from the UI). Why would you want to link that to another table from the public schema? Well, you can't manage RLS in the auth schema, so a user can only ever see its own data. So, if you want to be able to access users from your app, you need to create a public table, create a function and a trigger (like [that](https://github.com/orgs/supabase/discussions/306)).

Now, let's say that you want to dump your database and files, [you can](https://supabase.com/docs/guides/platform/migrating-within-supabase/backup-restore)! But wait, can you restore it just like that? Nope, you need some special things for that: use the provided script for files, and if it doesn't work when restoring, try [commenting out something](https://supabase.com/docs/guides/platform/migrating-within-supabase/backup-restore#important-project-restoration-notes) in the `schema.sql` file (by the way, here's our regex for that because that wasn't provided in the documentation: `ALTER\s+.*?\s+OWNER\s+TO\s+"supabase_admin"`).

Why is SMTP required by default? If you try to create a user, Supabase will try to send a confirmation email. But if you didn't set an SMTP provider (or equivalent), it will just results in an error, thus preventing you from creating a user. We had to use a dumb [mailcatcher](https://github.com/dockage/mailcatcher) just for that. And because, for some reason, turning the `confirm_email` to false [does not work](https://www.reddit.com/r/Supabase/comments/1ft1utu/turning_off_confirm_email_does_nothing/) (even with the Docker env variable), you can't disable that behaviour.

The last issue we had is with the compose stack: we couldn't use automated scripts, so we had to generate the compose file and just inserts it into Portainer. For this to work, you need to pre-create some files (a lot of them) that normally are in the Github repo. As someone who likes easy deployment and self-host as a hobby, I hate that.

What's great otherwise:
- seeding
- plugins
- edge functions
- CLI (for most parts)
- UI (for most parts)
- RLS

So yeah, I don't think that this is something production-ready yet. I haven't used the Cloud version, so I can't tell about that. Maybe I've misused Supabase and/or wanted to setup workflows that are not compatible with it yet, I don't know, but for what I've tested it doesn't feel right. If you use Supabase "the way it is supposed to be", you won't have any issue. But just deviate a little bit from it and it won't be fun.

**Reply by giosueDelgado** [reply]
Hi @dunivlorraine,

**Thank you very much for your timely and detailed feedback.**

I’m evaluating the use of Supabase for large clients to speed up backend development and I’m currently deciding between this and Strapi, even though they serve slightly different purposes. We will likely go with the self-hosted option as well (though for some large projects with millions of requests, we might consider the cloud). Seeing these limitations is a bit concerning.

Our goal is to have a scalable backend structure for a large multi-module ERP.
The main issue seems to be that it’s not possible to connect to a self-hosted environment from a local setup—can you confirm this?
If this aspect is resolved, could it be a good solution?

I’d love to hear the core team’s opinion on whether there’s a plan for this and on the other points—are they known issues or already resolved? 
FYI  @gc-ft 

Thanks, everyone!

**Reply by dunivlorraine** [reply]
> The main issue seems to be that it’s not possible to connect to a self-hosted environment from a local setup—can you confirm this?

Yes that's the main issue we had: you can't link a local database to a self-hosted one. But you can use some commands that target your self-hosted instance, it's just impractical, and for some operations, can't replace them.

> If this aspect is resolved, could it be a good solution?

TBH, I think it would become a good self-hosted solution. It would have its limitations (auth, SMTP, backups) but would really add a value to the development experience. I've also used [Pocketbase](https://pocketbase.io/) for another project, and overall I prefer their approach (control everything in the app and its tables, add custom functions that can interact directly with the DB/app, easy backups, etc.). But they are not built for the same purpose: Supabase can handle a huge number of requests and can be used as a multi-dimensional database thanks to its plugins (graph, timeseries, etc.), while Pocketbase is better suited for small projects and fast configuration.

**Reply by giosueDelgado** [reply]
Thanks for the clarification! In my case, I’m using Strapi for simpler projects, as it fits my needs well. However, I do think that better support for self-hosted environments in Supabase could make a big difference for many developers.

Is there someone from the Supabase team who could provide more insights on this? Are these known limitations, and is there any plan to address them in the future? It would be helpful to understand if there are improvements planned in this area.
